Fix VRFFluidCtrl cooling cycling issue by moving cycling ratio calc in compressor spd calc

yujiex opened the pull request:
Pull request overview
---------------------
 - Fixes #10105 
 - The range of the outdoor evaporating temperature is updated to match the valid range of the indoor unit evaporating temperature.

The following shows the relevant code flow chart. We observed a problem of non-matching minimum-speed compressor capacity at two places

![image](https://github.com/NREL/EnergyPlus/assets/13421161/69d5cf42-bb00-434b-8c8a-d4a5b95f8469)

For the same timestep, in the cycling ratio calculation, the demand is larger than minimum speed capacity, and the compressor is considered not cycling. However in the compressor speed calculation later on, the demand is smaller than the minimum speed capacity, and the compressor is considered cycling

![image](https://github.com/NREL/EnergyPlus/assets/13421161/a728cb92-0dce-430a-9cba-19ca58b00006)

The difference in the minimum-speed-capacity is caused by the difference in the evaporating temperature.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/d5e03e11-ebb9-4392-ac37-c8b0a2c7d33b)

To fix this, we propose to move the cycling calculation inside the compressor speed calculation as is shown in the following diagram.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/ed68ef72-fbe4-46d0-ba40-23b23cdcd17f)

### Effect of the fix

The following shows the output before the fix

<img width="1605" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/08b10c2c-d6bd-409c-9869-04b1d076d494">

The following shows the output after the fix. 

<img width="1608" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/4ebe412b-5c07-4ffb-a8ea-a2f155927146">

After the fix, runtime fractions are different now, where the lower-capacity one runs longer. For the same cooling rate, the heat pump cooling electricity rate and COP are more similar. The evaporating temperature output also reflected the following Te adjustment when the compressor is at the lowest speed (minimum loading index)

![image](https://github.com/NREL/EnergyPlus/assets/13421161/8b57a003-1fac-4558-a6b8-7edb72cd7532)
 

### Regression diffs

The regression diffs happen in the following files with VRF FluidTCtrl models
- US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF
- VariableRefrigerantFlow_FluidTCtrl_5Zone
- VariableRefrigerantFlow_FluidTCtrl_HR_5Zone
- VariableRefrigerantFlow_FluidTCtrl_wSuppHeater_5Zone

The four files all have table diffs in electricity rate. The following is in the table diff in  VariableRefrigerantFlow_FluidTCtrl_5Zone. These are anticipated as the cycling ratio calculation is modified and the cooling electricity rate should change.  As a result, the total electricity rate should also change.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/4c152cf5-6014-4ee9-a5e1-935a0d4d52d5)
![image](https://github.com/NREL/EnergyPlus/assets/13421161/6e90d796-ed97-4512-972d-dcb970ef9853)

Meter difference happens in the following two files. The differences are in Cooling:Electricity, Electricity:HVAC, and Electricity:Facility. The diffs are expected as cycling ratio can affect the cooling electricity rate, which leads to changes in these "Electricity:*" variables.
- VariableRefrigerantFlow_FluidTCtrl_5Zone
- VariableRefrigerantFlow_FluidTCtrl_HR_5Zone

The following is from the output of the "VariableRefrigerantFlow_FluidTCtrl_5Zone.idf".

![image](https://github.com/user-attachments/assets/fb1be51e-8dc1-4b46-9940-feed17193a06)

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
move cycling ratio calculation inside compressor spd calc

Commit:
move cycling into compressor speed calculation

Commit:
change outdoor unit evap temperature threshold

Commit:
overwrite indoor Te withoutdoor adjusted Te

Commit:
Merge remote-tracking branch 'origin/vrfFluidCtrlFixCoolingCycling' into vrfFluidCtrlFixCoolingCycling

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
overwrite IU Te, coil RTF update

Commit:
Only overwrite Coil RTF when in cooling mode

Commit:
ncomp calc need to multiply cycling ratio

yujiex commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -283,6 +283,25 @@ void SimulateVRF(EnergyPlusData &state,
+    } else {
+        PLF = 1.0;
+    }
+    if (coolingCoil.TotalCoolingEnergyRate > 0.0) {
```
this keeps the coil RTF up-to-date with the condenser side

yujiex commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14429,12 +14440,19 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
                 }
 
-                if (CapDiff > (Tolerance * Cap_Eva0)) NumIteCcap = 999;
+                if (CapDiff > (Tolerance * Cap_Eva0)) {
```
cycling ratio calculation is moved to here

rraustad replied:
This logic does not seem right to me. At line 14439 this same conditional is used. It says if the capacity difference is greater than some tolerance keep iterating. So when we get to line 14447, it's either due to max iterations or the capacity difference is less than tolerance. I guess if capacity difference is less than tolerance then we converged on the capacity that requires the compressor to run the entire time step. But if `(NumIteCcap >= 30)` then what does that mean? I'm not sure it means the compressor is cycling. I think it might mean the compressor is cycling if `Cap_Eva1 - Cap_Eva0` is positive but you don't know that since `CapDiff = std::abs(Cap_Eva1 - Cap_Eva0);` I would think that should be accounted for here. I'm not sure though since this is a very detailed model.

if (CapDiff > (Tolerance * Cap_Eva0) && (Cap_Eva1 - Cap_Eva0) >= 0.0)

yujiex replied:
Indeed. If Cap_Eva1 - Cap_Eva0 < 0 then it should not cycle.

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
update Te and RTF only happens in the FluidTCtrl model

mjwitte commented:
@yujiex For the defect file attached to #10105 these code changes don't alter the relative behavior between the two defect files (i.e. the stated speed at load range 1 has no impact on results other than the reported speed value, see https://github.com/NREL/EnergyPlus/issues/10105#issuecomment-1984151211). The results in both the 900 and 1320 file both change the same way.

But this PR does expose the fact that the initial check of load vs min-capacity has a significantly different value for `CompEvaporatingCAPSpdMin` than the equivalent check later on in the speed-selection pass. Looking at the code in develop, here's the first block in `VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl`:
https://github.com/NREL/EnergyPlus/blob/769463d9e9b6ec7059494f2db11ce8fa506b5a44/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L11531-L11543

And here's the second block:
https://github.com/NREL/EnergyPlus/blob/769463d9e9b6ec7059494f2db11ce8fa506b5a44/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L11838-L11842

So my question would be why the first one is using `CapMinTc = OutdoorDryBulb + this->SC` to evaluate the capacity curve, while the second block is using `Tdischarge` where
```
        Tdischarge = GetSatTemperatureRefrig(state, this->RefrigerantName, max(min(Pdischarge, RefPHigh), RefPLow), RefrigerantIndex, RoutineName);
```

yujiex commented:
> @yujiex For the defect file attached to #10105 these code changes don't alter the relative behavior between the two defect files (i.e. the stated speed at load range 1 has no impact on results other than the reported speed value, see [#10105 (comment)](https://github.com/NREL/EnergyPlus/issues/10105#issuecomment-1984151211)). The results in both the 900 and 1320 file both change the same way.
> 
> But this PR does expose the fact that the initial check of load vs min-capacity has a significantly different value for `CompEvaporatingCAPSpdMin` than the equivalent check later on in the speed-selection pass. Looking at the code in develop, here's the first block in `VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl`:
> 
> https://github.com/NREL/EnergyPlus/blob/769463d9e9b6ec7059494f2db11ce8fa506b5a44/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L11531-L11543
> 
> And here's the second block:
> 
> https://github.com/NREL/EnergyPlus/blob/769463d9e9b6ec7059494f2db11ce8fa506b5a44/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L11838-L11842
> 
> So my question would be why the first one is using `CapMinTc = OutdoorDryBulb + this->SC` to evaluate the capacity curve, while the second block is using `Tdischarge` where
> 
> ```
>         Tdischarge = GetSatTemperatureRefrig(state, this->RefrigerantName, max(min(Pdischarge, RefPHigh), RefPLow), RefrigerantIndex, RoutineName);
> ```

My understanding is the CompEvaporatingCAPSpdMin check in the first block before entering the compressor speed calculation might be a rougher guess and the second place considers the piping loss and is more accurate.

Commit:
test idf for different min speed level, will remove later

Commit:
account for cycling ratio in condenser heat release

Commit:
test case for VRFOU_CalcCompC low load cycling

mjwitte commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -283,6 +283,27 @@ void SimulateVRF(EnergyPlusData &state,
         if (state.dataHVACVarRefFlow->VRF(VRFCondenser).CondenserType == DataHeatBalance::RefrigCondenserType::Water)
             UpdateVRFCondenser(state, VRFCondenser);
     }
+    // yujie: update coil and IU evaporating temperature, also keep coil RTF updated with the condenser side cycling ratio, for the FluidTCtrl model
```
Current practice is to leave names out of comments.

yujiex replied:
I'll remove the name

mjwitte commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14429,12 +14442,20 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
 
-                Ncomp = this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction);
+                Ncomp = this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction) * CyclingRatio;
+                OUCondHeatRelease = Cap_Eva1 * CyclingRatio;
```
I don't think this is correct. Earlier `OUCondHeatRelease = TU_load + Pipe_Q + Ncomp;`
Possibly `OUCondHeatRelease  *= CyclingRatio;` would be correct here?

yujiex replied:
or would it be Ncomp + Cap_Eva1 * CyclingRatio? similar to the load > maximum capacity case
```
    if (CounterCompSpdTemp > NumOfCompSpdInput) {
        // Required load is beyond the maximum system capacity
        CompEvaporatingCAPSpd(NumOfCompSpdInput) =
            this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(NumOfCompSpdInput), T_discharge, T_suction);
        OUCondHeatRelease = Ncomp + CompEvaporatingCAPSpd(NumOfCompSpdInput);
        CompSpdActual = this->CompressorSpeed(NumOfCompSpdInput);
        Ncomp = CompEvaporatingPWRSpd(NumOfCompSpdInput);
    }
```

Ncomp is updated inside `VRFOU_CalcCompC`, so maybe that part needs to use the updated `Ncomp`?, When it is cycling,  `CyclingRatio = (TU_load + Pipe_Q) * C_cap_operation / Cap_Eva1`, then `Ncomp + Cap_Eva1 * CyclingRatio` would evaluates to `Ncomp + (TU_load + Pipe_Q) * C_cap_operation`, with the updated `Ncomp` and the `TU_load + Pipe_Q` considering the difference between the rated and real condition (`C_cap_operation`)

mjwitte replied:
That sounds ok. This function is very hard to follow, so I'll trust you on that.

rraustad replied:
It would be a nice test to plot some of the available report variables through a transition from CyclingRatio = 1 at some time step to CyclingRatio < 1 at future time steps to see if these report variables have a smooth transition from compressor on to compressor cycling. Whichever reports capture the equations being questioned here. Not sure why a component model is not reporting refrigerant pressures which could help in diagnostics. We should add refrigerant temperature/pressure reporting at some point.

yujiex replied:
good idea. I will produce some plot showing how CyclingRatio transition from 1 to < 1 and the other way

yujiex replied:
<img width="1030" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/c1281ed8-e5e0-4def-bfc6-c5c2a73a0f37">

[output change for cycling ratio from non-1 to 1.xlsx](https://github.com/NREL/EnergyPlus/files/14577311/output.change.for.cycling.ratio.from.non-1.to.1.xlsx)

I plotted two snapshot with cycling ratio changing from < 1 to 1 then back to < 1. The cooling electricity rate and condenser heat release seems to be smooth

dareumnam replied:
I'm curious why the VRF heat pump's cooling electricity rate is higher when the cycling ratio is less than 1 (for example, on 4/9 at 14:00 and 15:00) compared to when the cycling ratio is exactly 1 (on 4/9 at 16:00 and 17:00).

rraustad replied:
That's a good question. From [this figure](https://github.com/NREL/EnergyPlus/pull/10353#issuecomment-2110831149) suction temperature is also changing which will change the refrigerant mass flow rate at the same volumetric flow rate. So there are other things changing in this model. I guess I would add suction temperature and refrigerant mass flow rate to these plots to see if something can be identified as the driving force for the difference in cooling electricity rate. I also see 2 variables for OU fan power. I'm not sure what N_fan_OU_evap and m_air_evap are regarding the outdoor unit performance but I see that these variables are related to air mass flow rates. Are these variables also changing at these times? I think you will have to dig a little deeper into the model to understand if the Heat Pump Electricity Rate shown above is correct or not.

    // OU fan power
    N_fan_OU = N_fan_OU_evap + N_fan_OU_cond;

        N_fan_OU_evap = this->RatedOUFanPower * m_air_evap / m_air_rated;
        N_fan_OU_cond = this->RatedOUFanPower * m_air_cond / m_air_rated;

rraustad replied:
Oh, and add outdoor temperature since that is also a driving force of performance.

rraustad replied:
Also add VRF Heat Pump Outdoor Unit Evaporator Heat Extract Rate so we can see what is happening at the evaporator.

yujiex replied:
@dareumnam and @rraustad. I created the following plots. There is not an output variable for refrigerant flow. I will add some custom outputs later. The Outdoor Unit Evaporator Heat Extract Rate is 0 during cooling mode, in both periods, 04/06 and 04/09.

The following are a few plots for the 04/09 period. The fact that heat pump electricity is higher when cycling ratio < 1 than when cycling ratio = 1, seems to originate from the compressor power, which also has this behavior. I will dig deeper into this.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/a8e00572-81b7-41ad-b97a-7d27589e45d0)
[compare_outputs_cyclingRatio_non1to2_update.xlsx](https://github.com/user-attachments/files/15778642/compare_outputs_cyclingRatio_non1to2_update.xlsx)

yujiex replied:
I looked at the following two points, at 4/6 12:20 and 4/6/ 14:20. The former has heat pump cycling ratio = 0.949, the latter has cycling ratio of 1.0. The former has higher compressor power (which leads to higher heat pump electricity rate) than the latter. 
![image](https://github.com/NREL/EnergyPlus/assets/13421161/3262bd21-9ce7-45a3-aff0-81513a062cf3)

This is because its evaporating temperature is lower. The condensing temperature is pretty similar, lower evaporating temperature will lead to higher curve value (see the following compressor curve at this speed). 

![image](https://github.com/NREL/EnergyPlus/assets/13421161/2fcf98db-a453-4b6e-9e1f-61d54f6c6301)

As the compressor electricity rate is computed as the following, the compressor power (before multiplying cycling ratio is higher) at the 12:20 due to lower evaporating temperature.
```
Ncomp = this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction);
```

The evaporating temperature is lowered during cycling as there was a low-load modification mechanism in the fluid control VRF, where they enlarge the compressor evaporating capacity by changing the evaporating temperature (see the red shaded part), so that the equipment can cycle less.
![image](https://github.com/NREL/EnergyPlus/assets/13421161/268e4cf5-ac08-4a49-891b-bfd9f0edee99)

@rraustad @dareumnam

dareumnam replied:
Are the refrigerant mass flow rates the same at both points? (4/6 12:20 and 4/6/ 14:20)

yujiex replied:
Hi @dareumnam, there is not a refrigerant flow rate output variable. I used some debug prints to output this variable (`m_ref_IU_evap_i ` and `m_ref_IU_evap` in cooling mode). The refrigerant flow at 12:20 is 0.01058698505294939, and the refrigerant flow at 14:20 is 0.012904189061317716.

mjwitte commented on testfiles/speed_1320.idf:
```diff
@@ -0,0 +1,7429 @@
+!- ---------------------------------------------------------------------------------------------------------
+!- The Residential Prototype Building Models were developed by Pacific Northwest National Laboratory (PNNL),
```
The defect files got added by mistake.

yujiex replied:
let me remove it

mjwitte reviewed (commented):
A few more comments. I would like @rraustad to review also.

Commit:
OUCondHeatRelease = Ncomp + Cap_Eva1 * CyclingRatio;

Commit:
Revert "test idf for different min speed level, will remove later"

This reverts commit 96f9e2cd5f399276f7d961539537a446a2fe82eb.

Commit:
remove name tag in the comments

Commit:
fix unit test

Commit:
move cycling ratio after the two goto labels

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -283,6 +283,27 @@ void SimulateVRF(EnergyPlusData &state,
+        if (coolingCoil.TotalCoolingEnergyRate > 0.0) {
+            coolingCoil.CoolingCoilRuntimeFraction = state.dataHVACVarRefFlow->VRF(VRFCondenser).VRFCondCyclingRatio / PLF;
+        }
+    }
```
This is a strange placement of code in the Sim routine. Shouldn't any updates be completed in the Calc routine?

yujiex replied:
I just put this into the Calc routine.

Commit:
move IU Te and RTF updates into CalcVRFCondenser_FluidTCtrl

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -12572,6 +12558,25 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
+        }
+        if (coolingCoil.TotalCoolingEnergyRate > 0.0) {
+            coolingCoil.CoolingCoilRuntimeFraction = this->VRFCondCyclingRatio / PLF;
+        }
```
I don't see CoolingCoilRuntimeFraction getting set to 0 anywhere. So if TotalCoolingEnergyRate = 0, then CoolingCoilRuntimeFraction will be used from the last time step the compressor operated. I would set CoolingCoilRuntimeFraction = 0.0 in Init.

yujiex replied:
It is set to 0 in this function: CalcVRFCoolingCoil_FluidTCtrl
```
    // Initialize coil air side parameters
    CondInletTemp = 0.0;
    CondInletHumRat = 0.0;
    AirMassFlow = thisDXCoil.InletAirMassFlowRate;
    InletAirDryBulbTemp = thisDXCoil.InletAirTemp;
    InletAirEnthalpy = thisDXCoil.InletAirEnthalpy;
    InletAirHumRat = thisDXCoil.InletAirHumRat;
    state.dataHeatBal->HeatReclaimDXCoil(DXCoilNum).AvailCapacity = 0.0;
    thisDXCoil.CoolingCoilRuntimeFraction = 0.0;
```

rraustad replied:
Oh, I see. This is updating the coil model RTF, from the parent object. We should not do that. So why did you have to update the coil RTF? VRFCondCyclingRatio is already passed to SimDXCoil so the coil RTF should already be accurate, if that call is passing the updated cycling ratio.

yujiex replied:
The condenser side cycling ratio is updated in `CalcVRFCondenser_FluidTCtrl`. This took place after SimDXCoil, so the coil RTF is updated here to keep up with the updated condenser side cycling ratio

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14268,14 +14277,17 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
 
                 General::SolveRoot(state, 1.0e-3, MaxIter, SolFla, SmallLoadTe, f, MinOutdoorUnitTe,
-                                   T_suction);   // SmallLoadTe is the updated Te'
-                if (SolFla < 0) SmallLoadTe = 6; // MinOutdoorUnitTe; //SmallLoadTe( Te'_new ) is constant during iterations
```
This is poor logic for a SolveRoot failure. -1 means iteration limit exceeded and -2 means it can't find a solution. Using SolFla < 0 captures both and assumes there is a small load. Maybe this works, maybe it doesn't, I don't know. This is a separate issue, if this is actually an issue.

yujiex replied:
I will separate out SolFla=-1 and SolFla=-2 case. When it is -1 I will throw an error message of not reaching convergence. When it is -2, it would be the case where the load is small. 

In the -2 case, f(MinOutdoorUnitTe) and f(T_suction) can't be both positive as f(T_suction) > 0 would mean 
```
Q_evap_req * C_cap_operation / this->RatedEvapCapacity > CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction)
```
, which is 
```
Q_evap_req * C_cap_operation > this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction)
```
If this is the case, it shouldn't have been in `if (Q_evap_req * C_cap_operation <= CompEvaporatingCAPSpd(CounterCompSpdTemp)) {`

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14186,7 +14193,8 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
     Real64 Tfs;                            // Temperature of the air at the coil surface [C]]
-    Real64 Tolerance(0.05);                // Tolerance for condensing temperature calculation [C}
+    Real64 Tolerance(0.05);                // Tolerance for condensing temperature calculation [C]
+    Real64 TeTol(0.5);                     // Tolerance for the difference between Te and SmallLoadTe
```
These should be constexpr

yujiex replied:
changed to constexpr

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11719,6 +11703,8 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
 
         // Key outputs of this subroutine
+        Ncomp *= CyclingRatio;
+        Q_h_OU *= CyclingRatio;
```
Ncomp is power. Ncomp should multiply by RTF.

yujiex replied:
RTF was calculated later in the function. Should I follow this code and compute RTF here so that RTF is accessible here? Also I'm not sure why it is written like this instead of `VRFRTF = min(1.0, CyclingRatio)`

```
    VRFRTF = 0.0;
    // VRF Cooling and Heating Electric Power (output variables)
    if (this->OperatingMode == ModeCoolingOnly) {
        PartLoadFraction = 1.0;
        VRFRTF = min(1.0, (CyclingRatio / PartLoadFraction));

        this->ElecCoolingPower = state.dataHVACVarRefFlow->VRF(VRFCond).Ncomp + this->OUFanPower;
        this->ElecHeatingPower = 0;

    } else if (this->OperatingMode == ModeHeatingOnly) {
        PartLoadFraction = 1.0;
        VRFRTF = min(1.0, (CyclingRatio / PartLoadFraction));

        this->ElecCoolingPower = 0;
        this->ElecHeatingPower = this->Ncomp + this->OUFanPower;

    } else if (this->OperatingMode == ModeCoolingAndHeating) {
        PartLoadFraction = 1.0;
        VRFRTF = min(1.0, (CyclingRatio / PartLoadFraction));

        this->ElecCoolingPower = (this->Ncomp + this->OUFanPower) * this->IUEvapHeatRate / (this->IUCondHeatRate + this->IUEvapHeatRate);
        this->ElecHeatingPower = (this->Ncomp + this->OUFanPower) * this->IUCondHeatRate / (this->IUCondHeatRate + this->IUEvapHeatRate);

    } else {
        this->ElecCoolingPower = 0;
        this->ElecHeatingPower = 0;
    }
    this->VRFCondRTF = VRFRTF;
```

rraustad replied:
I don't know why it's written like this either. This is in the heat recovery code so maybe it is assumed the compressor runs the entire time step. Don't use this code without modifying `PartLoadFraction = 1.0;`. PLF needs to be calculated from the performance curve as a function of CyclingRatio as `PartLoadFraction = max(0.7, CurveValue(state, vrf.CoolPLFFPLR, CyclingRatio));`.

RTF = PLR/PLF. For VRF CyclingRatio is PLR so RTF = CyclingRatio/PLF.

yujiex replied:
In the component or physics-based VRF model, the OU doesn't have a curve for part load fraction. The available curves are the following
- Outdoor Unit Evaporating Temperature Function of Superheating Curve Name
- Outdoor Unit Condensing Temperature Function of Subcooling Curve Name
- Defrost Energy Input Ratio Modifier Function of Temperature Curve Name
- Loading Index i Evaporative Capacity Multiplier Function of Temperature Curve Name
- Loading Index i Compressor Power Multiplier Function of Temperature Curve Name

rraustad replied:
OK, so there is no cycling/start-up losses in this model. Using CyclingRatio is OK here.

yujiex replied:
right.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14429,12 +14442,20 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
+                }
 
                 Ncomp = this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction);
+                OUCondHeatRelease = Ncomp + Cap_Eva1;
```
Condenser heat rejection = power + evap capacity. How does this calculation relate to lines 11706 and 11707? Is OUCondHeatRelease the same as Q_h_OU at line 11707? And where is Q_c_OU *= cyclingRatio? I am getting confused as to which variables account for cycling and which don't. Does `OUCondHeatRelease = Ncomp + Cap_Eva1;` account for cycling?

yujiex replied:
Ncomp should probably account for cycling
OUCondHeatRelease was computed as the following in the `CounterCompSpdTemp > NumOfCompSpdInput case`
```
OUCondHeatRelease = Ncomp + CompEvaporatingCAPSpd(NumOfCompSpdInput);
```
So I basically followed this in the low-load approach. Originally I have the Ncomp and `OUCondHeatRelease` multiply by CyclingRatio (or maybe it should be RTF) in CalcCompH() function, but since the flow of the code is like 

```
        Ncomp = TU_CoolingLoad / this->CoolingCOP;
        Ncomp_new = Ncomp;
    Label10:;
        Q_h_OU = Q_c_TU_PL + Ncomp_new; // Ncomp_new may be updated during Iteration_Ncomp Label10

        ...

        // *VEF OU Compressor Simulation at cooling mode: Specify the compressor speed and power consumption
        this->VRFOU_CalcCompC(state,
                              TU_CoolingLoad,
                              Tsuction,
                              this->CondensingTemp,
                              Psuction,
                              T_comp_in,
                              h_comp_in,
                              h_IU_evap_in,
                              Pipe_Q_c,
                              CapMaxTc,
                              Q_h_OU,
                              CompSpdActual,
                              Ncomp,
                              CyclingRatio);

        if ((std::abs(Ncomp - Ncomp_new) > (Tolerance * Ncomp_new)) && (Counter < 30)) {
            Ncomp_new = Ncomp;
            Counter = Counter + 1;
            goto Label10;
        }
```
As the initial value of Ncomp (Ncomp = TU_CoolingLoad / this->CoolingCOP) does not have CyclingRatio in it, and the compressor is indeed going to cycling, then the cycling ratio will get multiplied multiple times, once each label10 loop. So I moved them to after the loop is completed.

```
OUCondHeatRelease = Ncomp + Cap_Eva1 = Ncomp + Cap_Eva0 / CyclingRatio;
```

```
OUCondHeatRelease * CyclingRatio = Ncomp * CyclingRatio + Cap_Eva0 = Ncomp * CyclingRatio + TU_load + Pipe_Q
```

rraustad replied:
The plots you make of a transition from CyclingRatio = 1 to CyclingRatio < 1 (later in a day when the cooling load is decreasing) will show if these equations are correct. Maybe start there to determine how these equations need to be organized.

yujiex replied:
The plot is in this comment https://github.com/NREL/EnergyPlus/pull/10353#discussion_r1521980123

dareumnam replied:
It doesn't seem like the cycling ratio is double-counted. If we want to make sure, maybe we could double-check the output result against the hand calculation result at a certain fixed timestep.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -14429,12 +14442,20 @@ void VRFCondenserEquipment::VRFOU_CalcCompC(EnergyPlusData &state,
-                if (CapDiff > (Tolerance * Cap_Eva0)) NumIteCcap = 999;
+                if (CapDiff > (Tolerance * Cap_Eva0)) {
+                    NumIteCcap = 999;
+                    CyclingRatio = (TU_load + Pipe_Q) * C_cap_operation / Cap_Eva1;
```
This math is confusing. (TU_load + Pipe_Q) is the load the compressor must meet. And Cap_Eva1 already includes CapFT. So shouldn't this be `(TU_load + Pipe_Q) / Cap_Eva1` ? or just `Cap_Eva0 / Cap_Eva1` ?

yujiex replied:
`Cap_Eva0 / Cap_Eva1` probably better

Commit:
make tolerance constexpr

Commit:
split -1 and -2 cases in SolveRoot in CalcCompC

Commit:
use Cap_Eva0 / Cap_Eva1 to compute CyclingRatio

Commit:
add Cap_Eva1 >= Cap_Eva0 check and fix unit test

nrel-bot-2c commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

yujiex commented:
@rraustad @mjwitte would you mind looking at this PR again? Thanks.

yujiex commented:
I examined the fix using the following two test idf files and the NY epw file in the weather folder.
[speed_900.idf.txt](https://github.com/NREL/EnergyPlus/files/15299905/speed_900.idf.txt)
[speed_1920.idf.txt](https://github.com/NREL/EnergyPlus/files/15299906/speed_1920.idf.txt)

The two idf files differ at the capacity multiplier curves at the lowest speed
![image](https://github.com/NREL/EnergyPlus/assets/13421161/18f7bdaa-4af2-4825-b16d-fd5b14607d51)

The following plot shows the cooling and power multiplier curves at three different temperature

![image](https://github.com/NREL/EnergyPlus/assets/13421161/c27b9ebd-3a1d-46bd-89fb-70684a89cfd0)

The plot on the left shows the cooling capacity curve divided by the power curves. The right one shows the distribution of the condensing and evaporating temperature. According to the distribution plot, the 30F and 35F curve 1 and curve 2 are more relevant at evaporating temperature near 10F (yellow highlighted). The plots later on are produced for the timestamps when the higher-capacity one is cycling (cycling ratio between 0 and 1).

![image](https://github.com/NREL/EnergyPlus/assets/13421161/426f44da-d483-4eb9-94d0-632a6af99366) 

In this branch, the cycling ratio, electricity rate, and COP are as follows, for a simulation period over July. x axis are indices of timesteps.
![image](https://github.com/NREL/EnergyPlus/assets/13421161/d35ef58a-7a21-4ccb-b39e-2f0a7e254f75)

![image](https://github.com/NREL/EnergyPlus/assets/13421161/2e6688bf-8e9a-40d8-a888-9c6ab14887b5)

![image](https://github.com/NREL/EnergyPlus/assets/13421161/21052756-4d19-4f2a-a2fb-eaa175d20869)


COP shouldn’t have such large discrepancy when the file using curve 1 is fully run and the other file is cycling
The spikes in the electricity rate plots are also not right. Investigating this issue now.

rraustad commented:
The VRF Heat Pump Cycling Ratio and Cooling Coil Total Cooling Rate figures make sense. The cycling ratio is higher with Curve 1 since there is a lower capacity multiplier and the coil provides the same cooling capacity to meet the same load. The other figures are harder to understand since there are 2 curves involved. It is also not clear at what speed the compressor is operating in these figures.

yujiex commented:
> The VRF Heat Pump Cycling Ratio and Cooling Coil Total Cooling Rate figures make sense. The cycling ratio is higher with Curve 1 since there is a lower capacity multiplier and the coil provides the same cooling capacity to meet the same load. The other figures are harder to understand since there are 2 curves involved. It is also not clear at what speed the compressor is operating in these figures.

I added a compressor speed plot at the bottom. It was using the "timestep" resolution.

I added a COP curve plot (cooling capacity curves divided by power curves).

rraustad commented:
The compressor speed is nearly constant in both cases? In order for the cooling capacity to change the suction pressure/temperature would have to change. And would look similar to the cooling rate plots. I don't see a minimum compressor speed in the inputs so I assume it's the Compressor Speed at Loading Index 1 (and if that's true then why did the Curve 2 compressor speed fall well below 1920?). I thought VRF systems modulated compressor speed, not suction conditions? I also thought that suction pressure and/or temperature was the control point in the VRF control algorithm and compressor speed would modulate to hold that condition constant. Maybe since this system is operating at the minimum speed that this is how it needs to be when the system operates. Can you plot suction pressure and temperature? I would expect the Curve 1 suction temperature to be constant at a minimum value when compressor speed increases above the minimum speed.

      A5 , \field Refrigerant Temperature Control Algorithm for Indoor Unit
           \type choice
           \key ConstantTemp
           \key VariableTemp
           \default VariableTemp
      N9 , \field Reference Evaporating Temperature for Indoor Unit
      N10, \field Reference Condensing Temperature for Indoor Unit
      N11, \field Variable Evaporating Temperature Minimum for Indoor Unit
      N12, \field Variable Evaporating Temperature Maximum for Indoor Unit
      N13, \field Variable Condensing Temperature Minimum for Indoor Unit
      N14, \field Variable Condensing Temperature Maximum for Indoor Unit

      N32, \field Compressor Speed at Loading Index 1
           \required-field
           \type real
           \begin-extensible
           \note Minimum compressor speed

yujiex commented:
> The compressor speed is nearly constant in both cases? In order for the cooling capacity to change the suction pressure/temperature would have to change. And would look similar to the cooling rate plots. I don't see a minimum compressor speed in the inputs so I assume it's the Compressor Speed at Loading Index 1 (and if that's true then why did the Curve 2 compressor speed fall well below 1920?). I thought VRF systems modulated compressor speed, not suction conditions? I also thought that suction pressure and/or temperature was the control point in the VRF control algorithm and compressor speed would modulate to hold that condition constant. Maybe since this system is operating at the minimum speed that this is how it needs to be when the system operates. Can you plot suction pressure and temperature? I would expect the Curve 1 suction temperature to be constant at a minimum value when compressor speed increases above the minimum speed.
> 
> ```
>   A5 , \field Refrigerant Temperature Control Algorithm for Indoor Unit
>        \type choice
>        \key ConstantTemp
>        \key VariableTemp
>        \default VariableTemp
>   N9 , \field Reference Evaporating Temperature for Indoor Unit
>   N10, \field Reference Condensing Temperature for Indoor Unit
>   N11, \field Variable Evaporating Temperature Minimum for Indoor Unit
>   N12, \field Variable Evaporating Temperature Maximum for Indoor Unit
>   N13, \field Variable Condensing Temperature Minimum for Indoor Unit
>   N14, \field Variable Condensing Temperature Maximum for Indoor Unit
> 
>   N32, \field Compressor Speed at Loading Index 1
>        \required-field
>        \type real
>        \begin-extensible
>        \note Minimum compressor speed
> ```

The compressor speed at the loading index 1 is the minimum speed of the VRF compressor. The VRF system will first modulate compressor speed to match the demand. If the demand is smaller than the minimum speed, then it starts to update the indoor unit evaporating temperature to further change the capacity to meet the low demand. If the lowest capacity after adjusting Te is still larger than the demand, it will start cycling.

These are the IU evaporating temperature and OU condensing temperature plot with the compressor speed, at detailed timestep resolution. It seems there are no pressure output variables.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/2b71e861-ca37-417d-b63e-dba959ff75d7)

yujiex commented:
I did some more investigation on the COP

I think the difference in COP and compressor electricity rate are expected as the cycling and low-speed-no-cycling uses different evaporating temperatures. 

I did the following investigation

We observed a large difference between compressor electricity rate (compressor power), and COP when one model is cycling and the other model is not cycling. The x-axis is the index of timestamps.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/69e99121-522b-4b9b-b80e-f84167d2d084)

I selected this time period for a closer look. During this period the compressor in the "curve 2" model is cycling and the compressor in the "curve 1" model is always running. Columns K and Q are the compressor power calculated using Te, Tc, and compressor power curves. Columns L and R are compressor power outputs from the simulation result file.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/d75bf0e4-f8a8-41bc-b63c-bb90b060ecee)

The following table and plot compare the evaporative capacity. Columns K and Q are the evaporative capacity calculated using Te, Tc, and evaporative capacity curves. Columns L and R are evaporative capacity outputs from the simulation result file.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/012ea6ce-9170-4132-a805-221c3655b49e)

I compared the COP in the following table and plot. Columns L and N are the COP calculated using Te, Tc, and COP curves. Columns L and R are COP outputs from the simulation result file. The COP columns are calculated using the corresponding evaporative capacity / (compressor power + OU fan power)

![image](https://github.com/NREL/EnergyPlus/assets/13421161/1c51bd49-80ef-4548-9628-2be8c6176ddf)

So I think the COP and Heat Pump Cooling Electricity rate difference are expected due to the change in the evaporating temperature.

The following is a excel with my calculation
[eplusout_comparing COP curve calculated and output.xlsx](https://github.com/NREL/EnergyPlus/files/15358285/eplusout_comparing.COP.curve.calculated.and.output.xlsx)

In the previous version of this comment, there were some noticeable differences between the simulated and calculated compressor power. It is because some of the timesteps the compressor is already above the minimum speed. The compressor power involves interpolation between both the first and second loading index curves. After considering this, the compressor power results between the calculated (in the excel) and the simulation output are really similar now. There are still some small differences in the evaporative capacity. This leads to some differences in the calculated and simulated COP

There are also some differences in the VRF heat pump's total cooling rate. This is likely due to the differences in the zone temperature.

![image](https://github.com/NREL/EnergyPlus/assets/13421161/23de17d1-27f0-4975-9aad-50e91acc971d)

yujiex commented:
@rraustad and @mjwitte . In the [comment above](https://github.com/NREL/EnergyPlus/pull/10353#issuecomment-2113118732), I added some more plots to verify the COP, cooling rate, and compressor power against the calculated results from using the curves. The evaporative capacity and power curves are in [this comment](https://github.com/NREL/EnergyPlus/pull/10353#issuecomment-2108794924). Please see if they make sense. Thanks.

dareumnam commented:
I believe @rraustad's insight on [Yujie's comment above](https://github.com/NREL/EnergyPlus/pull/10353#issuecomment-2113118732) is crucial for us to proceed with this PR. Any additional input would be greatly appreciated @rraustad.

dareumnam committed:
pull develop and merge conflict

rraustad commented:
This effort got a little off the rails with discussion. The original issue was a difference in power/COP with a change in the min speed curve limit. This is very hard to understand for a model that is changing coil air flow to meet a load, and the refrigerant side property (Te), which could then impact the condensing temp (Tc). The cycling issue with PLF should not change the energy use magnitude much because it's just a startup loss so comparing cycling to not cycling is just a small difference (more energy if cycling), if everything else is the same. But with this model, everything else is not the same. We should back up to the original issue and compare a develop to this branch result. Is energy use now as expected? Annual energy difference? Are the code changes reasonable?, etc.

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
Merge branch 'vrfFluidCtrlFixCoolingCycling' of https://github.com/NREL/EnergyPlus into vrfFluidCtrlFixCoolingCycling

yujiex commented:
> This effort got a little off the rails with discussion. The original issue was a difference in power/COP with a change in the min speed curve limit. This is very hard to understand for a model that is changing coil air flow to meet a load, and the refrigerant side property (Te), which could then impact the condensing temp (Tc). The cycling issue with PLF should not change the energy use magnitude much because it's just a startup loss so comparing cycling to not cycling is just a small difference (more energy if cycling), if everything else is the same. But with this model, everything else is not the same. We should back up to the original issue and compare a develop to this branch result. Is energy use now as expected? Annual energy difference? Are the code changes reasonable?, etc.

Regarding the original issue. The speed_900.idf has 8 speed levels, the lowest speed level curve is curve set 1. The speed_1920.idf has 7 speed levels, the lowest speed level curve is curve set 2. speed_1920.idf is derived from speed_900.idf by removing the lowest speed level curve set (there are two curves for each speed level). The following are the "Evaporative Capacity Multiplier Function of Temperature Curve Name" and "Compressor Power Multiplier Function of Temperature Curve Name" of Curve 1 and 2 at 35F are most relevant.
![image](https://github.com/NREL/EnergyPlus/assets/13421161/0ad85e6a-8d07-4985-bfe0-dc860a36f4e2)

The following is the develop branch output. We can see that both of them have cycling ratio=1. However, with different lowest-speed curves, they still deliver the same total cooling rate. This is unreasonable. 
![image](https://github.com/NREL/EnergyPlus/assets/13421161/6054fd97-86d3-4996-8562-d6262bd47b15)

The following is the feature branch output. We can see in both cases the VRF system is cycling. Since curve set 1 can supply lower cooling capacity for the same evaporating, condensing temperature combination, it should cycle less, which is indeed the case here. Compressor power for the speed_900 output is slightly larger than the speed_1920 one as it runs for longer period and that the power curve is higher.
![image](https://github.com/NREL/EnergyPlus/assets/13421161/0ec91041-960e-4ca2-8c7e-1c06ce0a1845)

Annual energy difference is shown in the following table
<img width="499" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/1275c48a-291a-44d9-a0dd-5fd81dc58cbb">

The following attaches the input idfs and output tables
[ouput_dev.zip](https://github.com/user-attachments/files/15974163/ouput_dev.zip)
[output_feature.zip](https://github.com/user-attachments/files/15974164/output_feature.zip)
[input.zip](https://github.com/user-attachments/files/15974169/input.zip)

@rraustad

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
remove temporary testfiles

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

yujiex commented:
@Myoldmopar  and @dareumnam would you mind taking another look at this PR? Thanks.

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

dareumnam commented:
So, in the original issue, the min-compressor-speed capacity was different at two points in the code flow. To fix this, you moved the cycling ratio calculation inside the compressor speed calculation. After the fix, the output from this branch shows a different cycling ratio, which seems reasonable. Could you please explain a bit more about whether the energy use or VRF cooling electricity rate are as expected? I would like to be convinced that the Heat Pump Electricity Rate shown above is correct.
@rraustad, do you have any further comments or suggestions?

yujiex commented:
> So, in the original issue, the min-compressor-speed capacity was different at two points in the code flow. To fix this, you moved the cycling ratio calculation inside the compressor speed calculation. After the fix, the output from this branch shows a different cycling ratio, which seems reasonable. Could you please explain a bit more about whether the energy use or VRF cooling electricity rate are as expected? I would like to be convinced that the Heat Pump Electricity Rate shown above is correct. @rraustad, do you have any further comments or suggestions?

@dareumnam Looking at the tables in this [comment](https://github.com/NREL/EnergyPlus/pull/10353#issuecomment-2189590504). Compressor power become much more similar after the fix. This is anticipated, as according to the power multiplier curve plots in the same comment, the curve 1 value is around 0.07 and the curve 2 value is around 0.1 (a 0.7 to 1 relationship). The curve value need to multiply the rated compressor power and cycling ratio to get the compressor power. The cycling ratio of the two example files are about 0.55 to 0.81 (see first row of the second table below the curve plots in that same comment), which is a 1 to 0.68 relationship. Since the power curve values are 0.7 to 1 (between the two examples) and the cycling ratios are 1 to 0.68, the final compressor power should be pretty similar.

This leads to very similar heat pump cooling electricity rate after the fix. This is reasonable as the cooling electricity rate is the sum of compressor power and fan power. Fan power here is the same. Compressor power is very similar as is explained above, heat pump cooling electricity rate should also be similar.

dareumnam commented:
Thanks for the detailed explanation which makes sense to me. I built this branch locally, and everything looks clean. The diffs are expected and well-explained. I appreciate your hard work on the VRF models, @yujiex. I believe this is ready to be merged. @Myoldmopar

yujiex commented:
Thanks @dareumnam

Commit:
change default value of MaxHeat(Cool)ingCapacity to MaxCap

Commit:
Revert "change default value of MaxHeat(Cool)ingCapacity to MaxCap"

This reverts commit 1c3523891c5123b77fb910b40ce6f5a56087912f.

Commit:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Commit:
fix unit test

Myoldmopar commented on src/EnergyPlus/Material.cc:
```diff
@@ -211,7 +211,7 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
     }
 
-    // yujie: This looks kind of silly, but we need it to keep the Materials array in IDF order.
+    // This looks kind of silly, but we need it to keep the Materials array in IDF order.
```
This caused a conflict.  I'll fix it.

Myoldmopar replied:
Actually maybe it wasn't this...?

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into vrfFluidCtrlFixCoolingCycling

Myoldmopar commented:
Alright, resolved the conflict and built/tested.  Everything is all happy.  This will merge assuming CI stays clean.

Myoldmopar commented:
Alright, merging here. Thanks @yujiex , and @dareumnam.

yujiex commented:
Thanks @Myoldmopar
